### PR TITLE
Topic/coregen

### PR DESCRIPTION
**Pull request by amosr**
/jury approved @jystic

### CODE REVIEW

**Comment by tranma** [comment]
did you try the pulling out sub-expressions as let bindings thing?

**Comment by amosr** [comment]
> did you try the pulling out sub-expressions as let bindings thing?

nah, but I did add priority to the bindings. I think this is generating some real programs, but it looks like there's an infinite loop in flatten simp

File: icicle-compiler/src/Icicle/Avalanche/Statement/Flatten/Exp.hs [hunk 1]
```diff
@@ -201,7 +201,7 @@ flatX a_fresh xx stm
              stm'   <- flatX'
                      ( xPrim (Core.PrimArray $ Core.PrimArrayMap tv tv')
                        `makeApps'` [upd, xVar n'vals] )
-                     (\v -> stm (mapPack tk tv (xVar n'keys) v))
+                     (\v -> stm (mapPack tk tv' (xVar n'keys) v))
```
Comment (amosr) [inline icicle-compiler/src/Icicle/Avalanche/Statement/Flatten/Exp.hs]: this is a real bug, uncovered by better test generation!

**Comment by amosr** [comment]
This PR makes the compiler tests a bit slower: 11:30 instead of 9:30 (all test suites)

File: icicle-compiler/src/Icicle/Avalanche/Statement/Flatten/Exp.hs (hunk 1 above)
Comment (tranma) [inline icicle-compiler/src/Icicle/Avalanche/Statement/Flatten/Exp.hs]: 😮😮😮

File: icicle-compiler/test/Icicle/Test/Arbitrary/Data.hs [hunk 2]
```diff
@@ -49,6 +53,22 @@ instance IsString Var where
...
 
+-- Sometimes it's nice to be able to pretty print our generated programs
+instance PP.Pretty Var where
+ pretty (Var t i) = PP.text (T.unpack t) <> PP.text (show i)
+
+-- Generate totally arbitrary, totally random variables
+instance Arbitrary Var where
+  arbitrary =
+   sized $ \size ->
+    Var <$> elements viruses <*> choose (0, size)
```
Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Arbitrary/Data.hs]: please use `cats`? `viruses` have something with a space in it, very annoying to read :(

Comment (amosr) [inline icicle-compiler/test/Icicle/Test/Arbitrary/Data.hs]: cats it is

**Comment by amosr** [comment]
Ok, I think this is ready to go. Some primitives are missing from the generators. I'll add them back in in following PRs.

* Latest/Buf primitives, the type and evaluation rules seem to be conflicting, so I've removed them from `genPrimMany`.
* Latest primitives should also be added back to `genPrimConstructor`.
* BufT not generated by `genType`.
* Arithmetic pow isn't generated because it doesn't work on Ints. It might as well be changed to only Double, since that's the only one that works right now. There are probably other arith prims missing too. (`genPrimMany`)
* Time primitives: do not generate PrimTimeMinusMonths. It doesn't take many months to underflow a date, which means then the evaluators don't agree. (This could be fixed by using a consistent time representation even when under/over flowing)
* Output types are restricted to those original PSV output code supported, but can these be relaxed now? (`genOutputType`)

! @jystic @tranma

**Comment by amosr** [comment]
also, it appears my "infinite loop" in simp flattened was just it being slow. which isn't great but ... I decreased the size of the programs.

**Comment by jystic** [comment]
👍 looks good to me, as discussed IRL we should put that list of things somewhere so we can keep track of where we're at

**Comment by amosr** [comment]
#616 issue for tracking missing things from generator

**Pull request merged**